presto-main/src/main/java/com/facebook/presto/execution/FailedQueryExecution.java

```diff
@@ -111,6 +111,12 @@ public void addStateChangeListener(StateChangeListener<QueryState> stateChangeLi
     }
 
     @Override
+    public void addQueryInfoStateChangeListener(StateChangeListener<QueryInfo> stateChangeListener)
+    {
+        stateChangeListener.stateChanged(queryInfo);
```

Comment:
This callback should be fired using a different thread.  This is an exiting bug in the `addStateChangeListener` code also.  Can you add a commit to fix this (the constructor already has an executor so this should be an easy fix)?